library/ssl_tls.c

```diff
@@ -6692,24 +6851,52 @@ size_t mbedtls_ssl_get_max_frag_len( const mbedtls_ssl_context *ssl )
 {
     size_t max_len;
 
-    /*
-     * Assume mfl_code is correct since it was checked when set
-     */
+    /* Assume mfl_code is correct since it was checked when set */
     max_len = ssl_mfl_code_to_length( ssl->conf->mfl_code );
 
-    /*
-     * Check if a smaller max length was negotiated
-     */
+    /* Check if a smaller max length was negotiated */
     if( ssl->session_out != NULL &&
         ssl_mfl_code_to_length( ssl->session_out->mfl_code ) < max_len )
     {
         max_len = ssl_mfl_code_to_length( ssl->session_out->mfl_code );
     }
 
-    return max_len;
+    /* During a handshake, use the value being negotiated */
+    if( ssl->session_negotiate != NULL &&
+        ssl_mfl_code_to_length( ssl->session_negotiate->mfl_code ) < max_len )
+    {
+        max_len = ssl_mfl_code_to_length( ssl->session_negotiate->mfl_code );
+    }
+
+    return( max_len );
 }
 #endif /* MBEDTLS_SSL_MAX_FRAGMENT_LENGTH */
 
+size_t mbedtls_ssl_get_max_record_payload( const mbedtls_ssl_context *ssl )
+{
+    size_t max_len = MBEDTLS_SSL_MAX_CONTENT_LEN;
+
+#if defined(MBEDTLS_SSL_MAX_FRAGMENT_LENGTH)
+    const size_t mfl = mbedtls_ssl_get_max_frag_len( ssl );
+
+    if( max_len > mfl )
+        max_len = mfl;
+#endif
+
+#if defined(MBEDTLS_SSL_PROTO_DTLS)
+    {
+        const size_t overhead = mbedtls_ssl_get_record_expansion( ssl );
```

Comment:
We must check the return code here; as far as I see, `mbedtls_ssl_get_record_expansion()` currently fails for ChaChaPoly suites, which means we are _ignoring_ the configured MTU entirely because `overhead` will be very large (as it was casted from a (small) negative error code to `size_t`.

Reply:
This is only a problem for renegotiations, though, because `mbedtls_ssl_get_record_expansion()` correctly returns `0` for the initial identity transform.